doc/hacking/style_guide.md

```diff
@@ -0,0 +1,187 @@
+   parameter list and anything else necessary for the function's signature.
+4. Opening brace on the same line, not its own line.
+5. Single empty line before next definition.
+
+For example:
+
+    /**
+     * String documenting the function, in JavaDoc style.
+     */
+    void ClassName::
+    function_name(int x, int y) const {
+      ...
+    }
+
+    /**
+     * ...
+     */
+    ...
+
+If the function is a constructor with an initializer list, the `:` goes on the
+same line as the constructor name, and the initializers are listed *one per
+line*, indented by two spaces.
+
+    ClassName::
+    ClassName() :
+      _a("a"),
+      _b("b"),
+      _c("c")
+    {
+    }
+
+Empty function bodies in a .cxx/.I file still contain a single line break (i.e.
+to put `}` on its own line).  In a .h file, an empty function can be given as
+`{}`
+
+Braces
+------
+
+Always put an opening `{` on the same line as the statement that calls for the
+brace.  There is exactly one exception to this: constructors with initializer
+lists.
+
+    SomeClass::
+    SomeClass() :
+      _one(1),
+      _two(2)
+    {
+      if (_one < _two) {
+        ...
+      }
+    }
+
+Comments
+--------
+
+Comments should be properly-formatted English, with proper capitalization and
+punctuation rules.  There are two spaces between sentences.
+
+Note that `//` is preferred over `/*` except in some special circumstances.
+
+Line limits and continuations
+-----------------------------
+
+Lines should be limited to 80 columns.  This is not a hard limit, but if a line
+does go over this length, consider a continuation.
+
+Continuation lines shall be aligned to the same column as appropriate, or
+indented two spaces otherwise:
+
+    if ((a &&
+         b) ||
+	c) {
+      ...
+    }
+
+When choosing where to break a line, these are the preferred places, from best
+to worst:
+
+1. If inside a comment, break the comment to as many lines as appropriate, but
+   only after moving the comment to its own line. (In other words, comments
+   long enough to need 
+2. After a comma in an argument/parameter list.
+3. After a binary operator (leaving the binary operator at the end of the line)
+4. In a function declaration, after the function's return type but before the
+   function's name. (Indented by two spaces.)
+5. In a log statement with `<<`, before a `<<`. (Avoid ending a line with `<<`)
+6. Around a string literal, at a natural word boundary, with a space left on
+   the previous line (e.g. `"This is a "` / `"continued string literal."`)
+7. After the `=` in an assignment operator.
+8. As a last resort, after the opening `(` in a function call.
+
+If none of these rules can apply cleanly, it is okay to have lines extend past
+the 80-column boundary.
+
+Alignment
+---------
+
+TBD, should discuss:
+
+1. Groups of declarations/definitions
+2. Groups of comments
+3. Pointers always being `obj *x;` and never `obj* x;`
```

Comment:
I realise this may not be the popular opinion, but I strongly lean towards the former, because this code isn't very obvious:
```
// Is only x a pointer, or are they both pointers?
Object* x, y;
```
It makes it look like the pointer is part of the type, and that x and y are both pointers, but that's not how C++ parses instances.  Putting the decorators where they belong makes it clear that only x is a pointer:
```
Object *x, y;
```

Reply:
It's also my personal choice of pointer style (I never knew that tidbit about the C++ parser) and what's overwhelmingly (universally?) used in the existing Panda codebase.

Reply:
@rdb I actually would prefer that if your variables have a different type, you declare them on seperate lines. 
```
Object *x;
Object y;
```

This makes it perfectly clear what type both variables have. I personally don't like declaring two lines in a single line at all, but I guess since we use it at many places, we should allow it.

Reply:
I would also like to discuss alignment of `*` and `&` operators. Should it be:
`int& x` or `int &x` ?  `int* x` or `int *x`?`

I think we had this discussion a while ago already. It depends on wheter you call it `x is a pointer to int` or `the type of x is an integer pointer`. I personally prefer the latter, but I think we use the former in the codebase.

Reply:
C++ types are best read right-to-left, for example:
```cpp
// foo is a reference to a const pointer to an integer
int *const &foo;
```

I would personally prefer aligning the `&` and `*` to the right, because it matches how C++ parses declarations, and because it matches what most of the codebase is currently doing.  That said, I think this should be a guideline, not a rule; I would not reject a PR simply because it happens to align the `*` and `&` on the other side.

I agree on declaring the variables on different lines if it could easily be confusing.  An exception is a simple case where there can be no confusion:
```cpp
int a, b, c;
```

Reply:
We could maybe set a preference for placement, and disallow declaring variables on the same line if they have a different type ?